Blish HUD/Controls/FormattedLabel.cs

```diff
@@ -0,0 +1,337 @@
+            if (_horizontalAlignment != HorizontalAlignment.Left) {
+                foreach (var item in _rectangles.GroupBy(x => x.Rectangle.Y)) {
+                    if (_horizontalAlignment == HorizontalAlignment.Center) {
+                        var combinedWidth = item.Sum(x => x.Rectangle.Width);
+                        var firstRectangleX = (Width / 2) - (combinedWidth / 2);
+
+                        var nextRectangleX = firstRectangleX;
+                        foreach (var rectangle in item) {
+                            rectangle.Rectangle.X = nextRectangleX;
+                            nextRectangleX = rectangle.Rectangle.X + rectangle.Rectangle.Width;
+                        }
+                    } else if (_horizontalAlignment == HorizontalAlignment.Right) {
+                        var reversedOrder = item.Reverse().ToArray();
+                        var nextRectangleX = Width - reversedOrder.First().Rectangle.Width;
+                        for (int i = 0; i < reversedOrder.Length; i++) {
+                            reversedOrder[i].Rectangle.X = nextRectangleX;
+
+                            if (i != reversedOrder.Length - 1) {
+                                nextRectangleX = reversedOrder[i].Rectangle.X - reversedOrder[i + 1].Rectangle.Width;
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        private void HandleVerticalAlignment() {
+            if (_verticalAlignment == VerticalAlignment.Middle) {
+                var yGroups = _rectangles.GroupBy(x => x.Rectangle.Y).ToArray();
+                var combinedHeight = yGroups.Select(x => x.OrderByDescending(x => x.Rectangle.Height).First().Rectangle.Height).Sum();
+                var firstRectangleY = (Height / 2) - (combinedHeight / 2);
+                var nextRectangleY = firstRectangleY;
+                foreach (var item in yGroups) {
+                    foreach (var rectangle in item) {
+                        rectangle.Rectangle.Y = nextRectangleY;
+                    }
+                    var maxHeightInRow = item.Max(x => x.Rectangle.Height);
+                    nextRectangleY += maxHeightInRow;
+                }
+            } else if (_verticalAlignment == VerticalAlignment.Bottom) {
+                var yGroups = _rectangles.GroupBy(x => x.Rectangle.Y).Reverse().ToArray();
+                var nextRectangleY = Height - yGroups.First().Max(x => x.Rectangle.Height);
+                foreach (var item in yGroups) {
+                    var maxHeightInRow = item.Max(x => x.Rectangle.Height);
+
+                    foreach (var rectangle in item) {
+                        rectangle.Rectangle.Y = nextRectangleY;
+                    }
+                    nextRectangleY -= maxHeightInRow;
+                }
+            }
+        }
+
+        public override void DoUpdate(GameTime gameTime) {
+            if (finishedInitialization) {
```

Comment:
Seems we could save some loops here by also checking that the mouse is over the control to begin with.  If it's not, there's no way it'd be inside of any of the inner rectangles.